### ISSUE TITLE
Beep when alarm triggered (and during arm_away delay) to remind user to enter code (or leave the building)

### ISSUE THREAD

**Issue by derekatkins** [thanksmister/android-mqtt-alarm-panel#132]
**Is your feature request related to a problem? Please describe.**
When I enter the house and the alarm pre-triggers, waiting for a code before the siren sounds, it would be nice if the alarm panel beeped to remind us to enter the code.

Similarly, when armed away, it would be nice for the panel to beep while we are during our exit time.

**Describe the solution you'd like**
When the alarm is triggered and the panel is asking for a code, the panel should beep every second or two to remind one to enter a code.  Similarly, it should beep during the delay of arm_away.

**Describe alternatives you've considered**
I have considered trying to get it to play an audio file of a beep, but I don't know how to get it to play it repeatedly, or to get it to stop once the code is entered.

**Additional context**
I am using 0.8.8.0 on a Fire 8.

**Comment by thanksmister** [comment]
This is already a feature, did you turn this on in the settings?   I am using in on a Fire 7 and it works fine. I uses whatever alarm sound you have on the device or alternatively a built-in beep sound file when the disarm alarm screen appears.

**Comment by derekatkins** [comment]
Thanks, @thanksmister !  I thought I looked through all the settings. I don't recall seeing it, but maybe I missed it, or maybe I didn't understand what the setting label meant.  I can't look at it right this minute but I'll take another look again for the setting and try it.
Are there any docs of the various settings and what they mean/do?  I'm also looking to set up certain screen "blanking" so that the pad in my bedroom will actually go dark when not-in-use (overnight).  I haven't played with all those settings, yet, so it too may be in there somewhere.
Thanks for what appears to be a great app!

**Comment by derekatkins** [comment]
Okay, got off my work call so I was able to try this.  I've got Settings -> Notification -> System Sounds turned on, and if I go to the System Settings -> Sound the alarm volume is pretty loud, but it doesn't beep when the alarm is activated (or if it does beep it only beeps once). 
Is this the correct setting?  Or is there another one in there?
Also, does it repeatedly beep the alarm or just fire off the sound once?

**Comment by thanksmister** [comment]
I do have a note about Fire Tablets: 

> There is a known issue with Fire OS devices from Amazon, they usually have a custom OS and may not include Daydream. You will then need to use the built-in screensaver features or you need to install an alternative solution. Some Fire OS devices also lack haptic feedback for key presses and the ability to customize the alarm sound.

Make sure this is on:

<img width="877" alt="Screen Shot 2020-08-17 at 12 17 01 PM" src="https://user-images.githubusercontent.com/142340/90412529-9ac28b80-e083-11ea-92f1-f3caac36bb29.png">

**Comment by thanksmister** [comment]
The beep occurs if the alarm has a delay when the mode is set to away or home.  The beep should be continuous when the keypad for entering the code is present.    This is not the same as "triggered" mode, that would be something you hook to an external siren.  Let me see if I can capture a video of mine.

**Comment by derekatkins** [comment]
Yes, I have "Notifications -> System Sounds" checked.

And...., just had an "aha" moment; there is no "delay" for armed_home.  Just set to armed_away and voila, when I open the door my pads were making noise.

Now the only time I'm missing noise is on the arming delay (e.g.. right after you arm_away before it times out).

**Comment by thanksmister** [comment]
There is no arming delay sound, only disarming, sorry!!

**Comment by derekatkins** [comment]
Any chance we can convert this to an RFE to add that feature?  :)

**Comment by thanksmister** [comment]
If you see my feature list, you would know that is unlikely.  I would be happy to have a PR for this feature, but it would also have to have a setting to turn on/off as well.

**Comment by derekatkins** [comment]
I'll see what I can do.  My android development Fu is low (and old and dated).  I do have the SDK from 2013 (android-sdk_r12-linux_x86.tgz and android-ndk-r9c-linux-x86_64.tar.bz2) but haven't done any android development in a very long time.

**Comment by thanksmister** [comment]
Well Kotlin is all the rage these days, so its going to be a little work.  I am migrating the project from Java/Kotlin to Flutter/Dart in the near future.  This is easier to develop with JavaScript/Dart and hopefully generates more contributors.  I will be doing this first with WallPanel as it has a better option to render LoveLace UI, I hope.

**Comment by derekatkins** [comment]
I'm old school.  I still remember when Java was *released* for the first time...  I remember when the C++ "compiler" was a pre-processor that would render C ;)
Still, I'll take a quick look and see if it'll be easy or not.  I can't imagine it would be a LOT of work to:
1) Add a preference, and
2) If that preference it set, call out to the alarm sound notifier while we're in the arming countdown.
But you never know.  I'm completely unfamiliar with the code-base :)

**Comment by thanksmister** [comment]
Well, if you put it like that sure, now you sound like my product manager.  How hard could it be.... haha.

**Comment by derekatkins** [comment]
It's just a SMOP, right?  ;)

**Comment by thanksmister** [comment]
Product things everything  is a SMOP, haha.

**Comment by derekatkins** [comment]
I just looked at the code for a bit...  I think I've found the 6 or so places I need to touch to go add a new preference.  So then it's just a question of how to use that preference. :)

But if you are going to rewrite the app, is it worth me spending time to make a PR?

**Comment by thanksmister** [comment]
I think we can just add the sound to the delay time for setting alarm to away, we can skip the preference.  Just use the same preference for system sounds that we use for entry sounds.  That should reduce the changes.   If you do not want to wait months, then it may be worth it.  I am slowing moving on new code base.

**Comment by derekatkins** [comment]
That, of course, is the harder problem.  (for me, at least).  It was relatively easy to start with the preference string and find | grep to figure out where it was declared, and then work backwards to determine how it was used.  Of course playing the sound is harder.
Right now I am trying to figure out:
1) where that dialog is instantiated, and
2) where to find the "play the alarm" code.

I think I found 2 (DialogUtils.kt:showAlarmDisableDialog), still tracking down 1.

**Comment by thanksmister** [comment]
The sound lives in the BaseAlarmView class and that is used for those dialogs.  If you look inAlarmDisableView class you can see how we check if its enabled, then  how we start and stop it. AlarmDisableView extends BaseAlarmView.

**Comment by thanksmister** [comment]
I think AlarmCodeView would need to be updated so if user set away mode and the count time timer was on, we would set the sound as we do in AlarmDisableView.

**Comment by thanksmister** [comment]
Well, maybe I am wrong, the peeping would be in the AlarmPendingView because you want the sound to happen when you see the count down timer on the main screen.  We need the view that displays on the main screen with the count down timer to include the sound methods as we use in BaseAlarmView and AlarmDisableView.

**Comment by derekatkins** [comment]
Thanks.  I'll go back and look later.  Of course, I don't know how to build and test my changes ;)

**Comment by thanksmister** [comment]
OH man.... you will need Android Studio and a lot of patience.

**Comment by derekatkins** [comment]
So imported and it says:
```
Build Config field cannot have a null parameter
```
and it opened up to this line of build.grade:
```
            buildConfigField 'Integer', 'ALARM_CODE', ALARM_CODE()
```

**Comment by derekatkins** [comment]
(The joys of having a Gigabit fiber connection -- downloading is fast)

**Comment by derekatkins** [comment]
AHA!  I needed to add an "alarmCode" to local.properties.  *sigh*
Is that documented anywhere?
Making progress on building...

**Comment by derekatkins** [comment]
Another build issue:  missing google-services.json file.
I was able to work around that by commenting out the "apply plugin: 'com.google.gms.google-services'" line in the build.gradle file.
After those two manual fixes, I was able to build the app.  Yay.
Now to look into testing it and then working on an update. :)

**Comment by derekatkins** [comment]
> Well, maybe I am wrong, the beeping would be in the AlarmPendingView because you want the sound to happen when you see the count down timer on the main screen. We need the view that displays on the main screen with the count down timer to include the sound methods as we use in BaseAlarmView and AlarmDisableView.

Yeah, this is definitely more complicated by the fact that AlarmPendingView does not derive from BaseAlarmView, but the sound code is in BaseAlarmView.  So I see two options:
1) refactor BaseAlarmView into a new superclass that provides the sound code and then derive both BaseAlarmView and AlarmPendingView from that new superclass, or
2) Copy-And-Paste the sound code into AlarmPendingView

The first option is probably the cleaner approach, but the second is definitely faster to implement and test.  :)

**Comment by thanksmister** [comment]
Copy the code or make the code until some util class used by both the main view and BaseAlarmView.

**Comment by derekatkins** [comment]
Fair enough.  I've copied it into AlarmPendingView for now.  
Just need to figure out the preference (and get it to stop playing with the timer is over).

**Comment by thanksmister** [comment]
Hmm, create a new public method in AlarmPendingView and pass in the preferences for hasSystemSound?

**Comment by derekatkins** [comment]
The issue is that the preference is passed in through like four or five levels of indirection from MainFragment where it is pulled from the configuration.

**Comment by derekatkins** [comment]
I've got it working except for the preference.  I'll work on that a bit more.

**Comment by thanksmister** [comment]
You are kotlin initiate now...

**Comment by derekatkins** [comment]
I think I got it..  The configuration is available in the ControlsFragment, which is where the AlarmPendingView is initialized, so I can probably set it from there before initializing.  Hold on..

**Comment by thanksmister** [comment]
Yeah, exactly, the parent views have the configuration injected, so they have reference, but views need public methods to pass in information.

**Comment by derekatkins** [comment]
w00t.  Got it.  I changed the pref and I got a different sound out.  Yay.
Now to actually fork and push the change to a PR for you!  :)

**Comment by thanksmister** [comment]
Great job!

**Comment by thanksmister** [comment]
Ok, merged and will do builds tomorrow!

**Comment by derekatkins** [comment]
Cool!  Thanks!

**Comment by thanksmister** [comment]
Released in v0.8.9.1
